Error converting RegularImmutableMap to SiteConfigResponse

With the following stack

```java
package appservice;

import java.util.ArrayList;
import io.pulumi.Stack;
import io.pulumi.azurenative.resources.ResourceGroup;
import io.pulumi.azurenative.resources.ResourceGroupArgs;
import io.pulumi.azurenative.web.AppServicePlan;
import io.pulumi.azurenative.web.AppServicePlanArgs;
import io.pulumi.azurenative.web.WebApp;
import io.pulumi.azurenative.web.WebAppArgs;
import io.pulumi.azurenative.web.inputs.NameValuePairArgs;
import io.pulumi.azurenative.web.inputs.SiteConfigArgs;
import io.pulumi.azurenative.web.inputs.SkuDescriptionArgs;
import io.pulumi.core.Output;
import io.pulumi.core.internal.annotations.OutputExport;
import io.pulumi.resources.CustomResourceOptions;

public final class MyStack extends Stack {

    @OutputExport(name="helloEndpoint", type=String.class, parameters={})
    private Output<String> helloEndpoint;

    public MyStack() {
        var resourceGroup = new ResourceGroup("resourceGroup",
                                              ResourceGroupArgs.builder().build(),
                                              CustomResourceOptions.Empty);

        var plan = new AppServicePlan("plan",
                AppServicePlanArgs.builder()
                        .setResourceGroupName(resourceGroup.getName().toInput())
                        .setKind("Linux")
                        .setReserved(true)
                        .setSku(SkuDescriptionArgs.builder().setName("B1").setTier("Basic").build())
                        .build(),
                CustomResourceOptions.Empty);

        var helloApp = new WebApp("helloApp",
                WebAppArgs.builder()
                        .setResourceGroupName(resourceGroup.getName().toInput())
                        .setServerFarmId(plan.getId().toInput())
                        .setSiteConfig(SiteConfigArgs.builder()
                                .setAppSettings(new ArrayList<>() {{
                                    add(NameValuePairArgs.builder()
                                            .setName("WEBSITES_ENABLE_APP_SERVICE_STORAGE")
                                            .setValue("false")
                                            .build());
                                }})
                                .setAlwaysOn(true)
                                .build())
                        .setHttpsOnly(true)
                        .build(),
                CustomResourceOptions.Empty);

        this.helloEndpoint = helloApp.getDefaultHostName()
            .applyValue(hostName -> String.format("https://%s/hello", hostName));
    }
}
```

Resources are deployed but Pulumi spits an error while converting WebApp's outputs:

```
java.lang.UnsupportedOperationException: Convert [io.pulumi.azurenative.web.WebApp.siteConfig]: Error converting 
'com.google.common.collect.RegularImmutableMap' to 'TypeShape{type=class 
io.pulumi.azurenative.web.outputs.SiteConfigResponse, parameters=[]}'. Expected type 
'io.pulumi.azurenative.web.outputs.SiteConfigResponse' (annotated with 
'io.pulumi.core.internal.annotations.OutputCustomType') to provide a constructor annotated with 
'io.pulumi.core.internal.annotations.OutputCustomType$Constructor', and the parameter names in the annotation matching 
the parameters being deserialized. Constructor 'private 
io.pulumi.azurenative.web.outputs.SiteConfigResponse(java.lang.Boolean,java.lang.String,java.lang.Boolean,io.pulumi.azurenative.web.outputs.ApiDefinitionInfoResponse,io.pulumi.azurenative.web.outputs.ApiManagementConfigResponse,java.lang.String,java.util.List,java.lang.Boolean,io.pulumi.azurenative.web.outputs.AutoHealRulesResponse,java.lang.String,java.util.Map,java.util.List,io.pulumi.azurenative.web.outputs.CorsSettingsResponse,java.util.List,java.lang.Boolean,java.lang.String,io.pulumi.azurenative.web.outputs.ExperimentsResponse,java.lang.String,java.lang.Integer,java.lang.Boolean,java.util.List,java.lang.String,java.lang.Boolean,java.lang.Boolean,java.util.List,java.lang.String,java.lang.String,java.lang.String,java.lang.String,io.pulumi.azurenative.web.outputs.SiteLimitsResponse,java.lang.String,java.lang.String,java.lang.Boolean,java.lang.Integer,io.pulumi.azurenative.web.outputs.SiteMachineKeyResponse,java.lang.String,java.lang.Integer,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,io.pulumi.azurenative.web.outputs.PushSettingsResponse,java.lang.String,java.lang.Boolean,java.lang.String,java.lang.Boolean,java.lang.String,java.util.List,java.lang.Boolean,java.lang.String,java.lang.String,java.lang.String,java.lang.Boolean,java.util.List,java.lang.String,java.lang.Integer,java.lang.Boolean,java.lang.Boolean,java.lang.String,java.lang.String,java.lang.Integer)' parameter named 'machineKey' (nr 34 starting from 0) lacks @javax.annotation.Nullable annotation, so the value is required, but there is no value to deserialize.
    	at io.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:95)
    	at io.pulumi.serialization.internal.Converter.convertValue(Converter.java:67)
    	at io.pulumi.deployment.internal.DeploymentImpl$ReadOrRegisterResource.lambda$completeResourceAsync$0(DeploymentImpl.java:1074)
    	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
    	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
    	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
    	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
    	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
    	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
    	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
    Caused by: java.lang.IllegalStateException: Expected type 'io.pulumi.azurenative.web.outputs.SiteConfigResponse' (annotated with 'io.pulumi.core.internal.annotations.OutputCustomType') to provide a constructor annotated with 'io.pulumi.core.internal.annotations.OutputCustomType$Constructor', and the parameter names in the annotation matching the parameters being deserialized. Constructor 'private io.pulumi.azurenative.web.outputs.SiteConfigResponse(java.lang.Boolean,java.lang.String,java.lang.Boolean,io.pulumi.azurenative.web.outputs.ApiDefinitionInfoResponse,io.pulumi.azurenative.web.outputs.ApiManagementConfigResponse,java.lang.String,java.util.List,java.lang.Boolean,io.pulumi.azurenative.web.outputs.AutoHealRulesResponse,java.lang.String,java.util.Map,java.util.List,io.pulumi.azurenative.web.outputs.CorsSettingsResponse,java.util.List,java.lang.Boolean,java.lang.String,io.pulumi.azurenative.web.outputs.ExperimentsResponse,java.lang.String,java.lang.Integer,java.lang.Boolean,java.util.List,java.lang.String,java.lang.Boolean,java.lang.Boolean,java.util.List,java.lang.String,java.lang.String,java.lang.String,java.lang.String,io.pulumi.azurenative.web.outputs.SiteLimitsResponse,java.lang.String,java.lang.String,java.lang.Boolean,java.lang.Integer,io.pulumi.azurenative.web.outputs.SiteMachineKeyResponse,java.lang.String,java.lang.Integer,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,java.lang.Integer,java.lang.String,java.lang.String,io.pulumi.azurenative.web.outputs.PushSettingsResponse,java.lang.String,java.lang.Boolean,java.lang.String,java.lang.Boolean,java.lang.String,java.util.List,java.lang.Boolean,java.lang.String,java.lang.String,java.lang.String,java.lang.Boolean,java.util.List,java.lang.String,java.lang.Integer,java.lang.Boolean,java.lang.Boolean,java.lang.String,java.lang.String,java.lang.Integer)' parameter named 'machineKey' (nr 34 starting from 0) lacks @javax.annotation.Nullable annotation, so the value is required, but there is no value to deserialize.
    	at io.pulumi.serialization.internal.Converter.tryConvertObjectInner(Converter.java:335)
    	at io.pulumi.serialization.internal.Converter.convertObjectUntyped(Converter.java:91)
```

--- a/sdk/jvm/pulumi/src/main/java/io/pulumi/core/internal/OutputCompletionSource.java
+++ b/sdk/jvm/pulumi/src/main/java/io/pulumi/core/internal/OutputCompletionSource.java
@@ -80,8 +80,10 @@ public class OutputCompletionSource<T> {
         ));
     }
 
-    public void setValue(String context, Value value, ImmutableSet<Resource> depsOrEmpty) {
-        setValue(Converter.convertValue(
+    public void setValue(Converter converter, String context, Value value, ImmutableSet<Resource> depsOrEmpty) {
+        // we need to call the converter here, inside this class where we know the T,
+        // the T type will get lost at the call site due to Java generics limitations
+        setValue(converter.convertValue(
                 context,
                 value,
                 getTypeShape(),

--- a/sdk/jvm/pulumi/src/main/java/io/pulumi/deployment/internal/DeploymentImpl.java
+++ b/sdk/jvm/pulumi/src/main/java/io/pulumi/deployment/internal/DeploymentImpl.java
@@ -68,6 +68,7 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
     private final Log log;
     private final FeatureSupport featureSupport;
     private final PropertiesSerializer serialization;
+    private final Converter converter;
     private final Invoke invoke;
     private final Call call;
     private final Prepare prepare;
@@ -92,13 +93,14 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
         this.log = new Log(state.logger, DeploymentState.ExcessiveDebugOutput);
         this.featureSupport = new FeatureSupport(state.monitor);
         this.serialization = new PropertiesSerializer(this.log);
-        this.invoke = new Invoke(this.log, state.monitor, this.featureSupport, this.serialization);
+        this.converter = new Converter(this.log);
+        this.invoke = new Invoke(this.log, state.monitor, this.featureSupport, this.serialization, this.converter);
         this.rootResource = new RootResource(state.engine);
         this.prepare = new Prepare(this.log, this.featureSupport, this.rootResource, this.serialization);
-        this.call = new Call(this.log, state.monitor, this.prepare, this.serialization);
+        this.call = new Call(this.log, state.monitor, this.prepare, this.serialization, this.converter);
         this.readResource = new ReadResource(this.log, this.prepare, state.monitor);
         this.registerResource = new RegisterResource(this.log, this.prepare, state.monitor);
-        this.readOrRegisterResource = new ReadOrRegisterResource(state.runner, this.invoke, this.readResource, this.registerResource, state.isDryRun);
+        this.readOrRegisterResource = new ReadOrRegisterResource(state.runner, this.invoke, this.readResource, this.registerResource, this.converter, state.isDryRun);
         this.registerResourceOutputs = new RegisterResourceOutputs(this.log, state.runner, state.monitor, this.featureSupport, this.serialization);
     }
 
@@ -402,12 +404,20 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
         private final Monitor monitor;
         private final FeatureSupport featureSupport;
         private final PropertiesSerializer serialization;
+        private final Converter converter;
 
-        private Invoke(Log log, Monitor monitor, FeatureSupport featureSupport, PropertiesSerializer serialization) {
+        private Invoke(
+                Log log,
+                Monitor monitor,
+                FeatureSupport featureSupport,
+                PropertiesSerializer serialization,
+                Converter converter
+        ) {
             this.log = Objects.requireNonNull(log);
             this.monitor = Objects.requireNonNull(monitor);
             this.featureSupport = Objects.requireNonNull(featureSupport);
             this.serialization = Objects.requireNonNull(serialization);
+            this.converter = Objects.requireNonNull(converter);
         }
 
         public <T> Output<T> invoke(String token, TypeShape<T> targetType, InvokeArgs args) {
@@ -458,7 +468,7 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
             // `invoke`.
 
             return invokeRawAsync(token, args, options)
-                    .thenApply(result -> Converter.convertValue(
+                    .thenApply(result -> this.converter.convertValue(
                             String.format("%s result", token),
                             Value.newBuilder()
                                     .setStructValue(result.serialized)
@@ -484,7 +494,7 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
 
         public <T> CompletableFuture<T> invokeAsync(String token, TypeShape<T> targetType, InvokeArgs args, InvokeOptions options) {
             return invokeRawAsync(token, args, options).thenApply(
-                    result -> Converter.convertValue(
+                    result -> this.converter.convertValue(
                             String.format("%s result", token),
                             Value.newBuilder()
                                     .setStructValue(result.serialized)
@@ -587,12 +597,20 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
         private final Monitor monitor;
         private final Prepare prepare;
         private final PropertiesSerializer serialization;
+        private final Converter converter;
 
-        public Call(Log log, Monitor monitor, Prepare prepare, PropertiesSerializer serialization) {
+        public Call(
+                Log log,
+                Monitor monitor,
+                Prepare prepare,
+                PropertiesSerializer serialization,
+                Converter converter
+        ) {
             this.log = Objects.requireNonNull(log);
             this.monitor = Objects.requireNonNull(monitor);
             this.prepare = Objects.requireNonNull(prepare);
             this.serialization = Objects.requireNonNull(serialization);
+            this.converter = Objects.requireNonNull(converter);
         }
 
         void call(String token, CallArgs args) {
@@ -626,7 +644,7 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
             Objects.requireNonNull(options);
 
             return callRawAsync(token, args, self, options).thenApply(
-                    r -> Converter.convertValue(
+                    r -> this.converter.convertValue(
                             String.format("%s result", token),
                             Value.newBuilder()
                                     .setStructValue(r.result)
@@ -1004,13 +1022,22 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
         private final Invoke invoke;
         private final ReadResource readResource;
         private final RegisterResource registerResource;
+        private final Converter converter;
         private final boolean isDryRun;
 
-        private ReadOrRegisterResource(Runner runner, Invoke invoke, ReadResource readResource, RegisterResource registerResource, boolean isDryRun) {
+        private ReadOrRegisterResource(
+                Runner runner,
+                Invoke invoke,
+                ReadResource readResource,
+                RegisterResource registerResource,
+                Converter converter,
+                boolean isDryRun
+        ) {
             this.runner = Objects.requireNonNull(runner);
             this.invoke = Objects.requireNonNull(invoke);
             this.readResource = Objects.requireNonNull(readResource);
             this.registerResource = Objects.requireNonNull(registerResource);
+            this.converter = Objects.requireNonNull(converter);
             this.isDryRun = isDryRun;
         }
 
@@ -1082,6 +1109,7 @@ public class DeploymentImpl extends DeploymentInstanceHolder implements Deployme
                                             var contextInfo = String.format("%s.%s", resource.getClass().getTypeName(), fieldName);
                                             var depsOrEmpty = Maps.tryGetValue(dependencies, fieldName).orElse(ImmutableSet.of());
                                             completionSource.setValue(
+                                                    this.converter,
                                                     contextInfo,
                                                     value.get(),
                                                     depsOrEmpty

--- a/sdk/jvm/pulumi/src/main/java/io/pulumi/deployment/internal/DeploymentTests.java
+++ b/sdk/jvm/pulumi/src/main/java/io/pulumi/deployment/internal/DeploymentTests.java
@@ -11,6 +11,8 @@ import io.pulumi.deployment.Mocks;
 import io.pulumi.deployment.internal.DeploymentImpl.DefaultEngineLogger;
 import io.pulumi.exceptions.RunException;
 import io.pulumi.resources.Resource;
+import org.mockito.ArgumentMatchers;
+import org.mockito.Mockito;
 
 import javax.annotation.Nullable;
 import java.util.Map;
@@ -23,8 +25,6 @@ import java.util.stream.Collectors;
 
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
-import org.mockito.ArgumentMatchers;
-import org.mockito.Mockito;
 
 public class DeploymentTests {
 
@@ -332,6 +332,10 @@ public class DeploymentTests {
 
     public static Log mockLog() {
         return mockLog(defaultLogger(), () -> Mockito.mock(Engine.class));
+    }
+
+    public static Log mockLog(Logger logger) {
+        return mockLog(logger, () -> Mockito.mock(Engine.class));
     }
 
     public static Log mockLog(Logger logger, Supplier<Engine> engine) {

--- a/sdk/jvm/pulumi/src/main/java/io/pulumi/serialization/internal/Converter.java
+++ b/sdk/jvm/pulumi/src/main/java/io/pulumi/serialization/internal/Converter.java
@@ -8,6 +8,7 @@ import com.google.gson.JsonElement;
 import com.google.gson.JsonNull;
 import com.google.gson.stream.JsonWriter;
 import com.google.protobuf.Value;
+import io.pulumi.Log;
 import io.pulumi.core.Archive;
 import io.pulumi.core.Archive.InvalidArchive;
 import io.pulumi.core.Asset.InvalidAsset;
@@ -29,6 +30,7 @@ import java.lang.reflect.Parameter;
 import java.util.*;
 import java.util.function.Function;
 
+import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.joining;
 
 /**
@@ -37,25 +39,27 @@ import static java.util.stream.Collectors.joining;
 @ParametersAreNonnullByDefault
 public class Converter {
 
-    private Converter() {
-        throw new UnsupportedOperationException("static class");
+    private final Log log;
+
+    public Converter(Log log) {
+        this.log = requireNonNull(log);
     }
 
-    public static <T> InputOutputData<T> convertValue(String context, Value value, Class<T> targetType) {
+    public <T> InputOutputData<T> convertValue(String context, Value value, Class<T> targetType) {
         return convertValue(context, value, TypeShape.of(targetType));
     }
 
-    public static <T> InputOutputData<T> convertValue(String context, Value value, TypeShape<T> targetType) {
+    public <T> InputOutputData<T> convertValue(String context, Value value, TypeShape<T> targetType) {
         return convertValue(context, value, targetType, ImmutableSet.of());
     }
 
-    public static <T> InputOutputData<T> convertValue(
+    public <T> InputOutputData<T> convertValue(
             String context, Value value, TypeShape<T> targetType, ImmutableSet<Resource> resources
     ) {
-        Objects.requireNonNull(context);
-        Objects.requireNonNull(value);
-        Objects.requireNonNull(targetType);
-        Objects.requireNonNull(resources);
+        requireNonNull(context);
+        requireNonNull(value);
+        requireNonNull(targetType);
+        requireNonNull(resources);
 
         checkTargetType(context, targetType);
 
@@ -86,7 +90,7 @@ public class Converter {
     }
 
     @Nullable
-    private static Object convertObjectUntyped(String context, @Nullable Object value, TypeShape<?> targetType) {
+    private Object convertObjectUntyped(String context, @Nullable Object value, TypeShape<?> targetType) {
         try {
             return tryConvertObjectInner(context, value, targetType);
         } catch (UnsupportedOperationException ex) {
@@ -103,7 +107,7 @@ public class Converter {
     }
 
     @Nullable
-    private static Object tryConvertObjectInner(
+    private Object tryConvertObjectInner(
             String context, @Nullable Object value, TypeShape<?> targetType
     ) {
         var targetIsOptional = Optional.class.isAssignableFrom(targetType.getType());
@@ -329,22 +333,23 @@ public class Converter {
                             argValue,
                             TypeShape.extract(parameter)
                     );
-                } else if (parameter.isAnnotationPresent(Nullable.class)) {
-                    arguments[i] = null;
                 } else {
-                    throw new IllegalStateException(String.format(
-                            "Expected type '%s' (annotated with '%s') to provide a constructor annotated with '%s', " +
-                                    "and the parameter names in the annotation matching the parameters being deserialized. " +
-                                    "Constructor '%s' parameter named '%s' (nr %d starting from 0) lacks @%s annotation, " +
-                                    "so the value is required, but there is no value to deserialize.",
-                            targetType.getTypeName(),
-                            OutputCustomType.class.getTypeName(),
-                            OutputCustomType.Constructor.class.getTypeName(),
-                            constructor,
-                            parameterName,
-                            i,
-                            Nullable.class.getTypeName()
-                    ));
+                    arguments[i] = null;
+                    if (!parameter.isAnnotationPresent(Nullable.class)) {
+                        log.debug(String.format(
+                                "Expected type '%s' (annotated with '%s') to provide a constructor annotated with '%s', " +
+                                        "and the parameter names in the annotation matching the parameters being deserialized. " +
+                                        "Constructor '%s' parameter named '%s' (nr %d starting from 0) lacks @%s annotation, " +
+                                        "so the value is required, but there is no value to deserialize.",
+                                targetType.getTypeName(),
+                                OutputCustomType.class.getTypeName(),
+                                OutputCustomType.Constructor.class.getTypeName(),
+                                constructor,
+                                parameterName,
+                                i,
+                                Nullable.class.getTypeName()
+                        ));
+                    }
                 }
             }
 
@@ -364,7 +369,7 @@ public class Converter {
         }
     }
 
-    private static JsonElement tryConvertJsonElement(String context, Object value) {
+    private JsonElement tryConvertJsonElement(String context, Object value) {
         var gson = new Gson();
         StringWriter stringWriter = new StringWriter();
         try {
@@ -376,7 +381,7 @@ public class Converter {
         return gson.fromJson(stringWriter.toString(), JsonElement.class);
     }
 
-    private static void tryWriteJson(String context, JsonWriter jsonWriter, @Nullable Object value) throws IOException {
+    private void tryWriteJson(String context, JsonWriter jsonWriter, @Nullable Object value) throws IOException {
         if (value == null) {
             jsonWriter.nullValue();
             return;
@@ -423,7 +428,7 @@ public class Converter {
         ));
     }
 
-    private static <T> T tryEnsureType(String context, Object value, TypeShape<T> targetType) {
+    private <T> T tryEnsureType(String context, Object value, TypeShape<T> targetType) {
         if (targetType.getType().isInstance(value)
                 || (boolean.class.isAssignableFrom(targetType.getType()) && value instanceof Boolean)
                 || (double.class.isAssignableFrom(targetType.getType()) && value instanceof Double)
@@ -439,7 +444,7 @@ public class Converter {
         }
     }
 
-    private static Either<Object, Object> tryConvertOneOf(String context, Object value, TypeShape<?> targetType) {
+    private Either<Object, Object> tryConvertOneOf(String context, Object value, TypeShape<?> targetType) {
         var leftType = targetType.getParameter(0)
                 .orElseThrow(() -> new IllegalStateException("Expected a left parameter type for the Either, got none"));
         var rightType = targetType.getParameter(1)
@@ -471,7 +476,7 @@ public class Converter {
         }
     }
 
-    private static ImmutableList<Object> tryConvertList(String context, Object value, TypeShape<?> targetType) {
+    private ImmutableList<Object> tryConvertList(String context, Object value, TypeShape<?> targetType) {
         if (!List.class.isAssignableFrom(value.getClass())) {
             throw new IllegalArgumentException(String.format(
                     "%s; Expected List but got '%s' while deserializing", context, value.getClass().getTypeName()
@@ -493,7 +498,7 @@ public class Converter {
         return builder.build();
     }
 
-    private static ImmutableMap<String, Object> tryConvertMap(String context, Object value, TypeShape<?> targetType) {
+    private ImmutableMap<String, Object> tryConvertMap(String context, Object value, TypeShape<?> targetType) {
         if (!Map.class.isAssignableFrom(value.getClass())) {
             throw new IllegalArgumentException(String.format(
                     "%s; Expected Map but got '%s' while deserializing", context, value.getClass().getTypeName()
@@ -516,14 +521,12 @@ public class Converter {
         return builder.build();
     }
 
-    // TODO
-
-    public static void checkTargetType(String context, TypeShape<?> targetType) {
+    public void checkTargetType(String context, TypeShape<?> targetType) {
         checkTargetType(context, targetType, new HashSet<>());
     }
 
     // pre-check for performance reasons
-    public static void checkTargetType(String context, TypeShape<?> targetType, HashSet<Class<?>> seenTypes) {
+    public void checkTargetType(String context, TypeShape<?> targetType, HashSet<Class<?>> seenTypes) {
 
         // types can be recursive.  So only dive into a type if it's the first time we're seeing it.
         if (!seenTypes.add(targetType.getType())) {

--- a/sdk/jvm/pulumi/src/test/java/io/pulumi/serialization/internal/ComplexTypeConverterTest.java
+++ b/sdk/jvm/pulumi/src/test/java/io/pulumi/serialization/internal/ComplexTypeConverterTest.java
@@ -2,20 +2,28 @@ package io.pulumi.serialization.internal;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
+import io.pulumi.Log;
 import io.pulumi.core.annotations.OutputCustomType;
+import io.pulumi.deployment.internal.DeploymentTests;
+import io.pulumi.deployment.internal.InMemoryLogger;
 import io.pulumi.serialization.internal.ConverterTests.ContainerSize;
 import org.junit.jupiter.api.Test;
 
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.logging.Level;
 
 import static io.pulumi.serialization.internal.ConverterTests.ContainerColor;
 import static io.pulumi.serialization.internal.ConverterTests.ContainerColor.Blue;
 import static io.pulumi.serialization.internal.ConverterTests.serializeToValueAsync;
+import static io.pulumi.test.internal.assertj.PulumiConditions.containsString;
 import static org.assertj.core.api.Assertions.assertThat;
 
 class ComplexTypeConverterTest {
+
+    private final static Log log = DeploymentTests.mockLog();
 
     @OutputCustomType
     public static class ComplexType1 {
@@ -49,6 +57,7 @@ class ComplexTypeConverterTest {
 
     @Test
     void testTestComplexType1() {
+        var converter = new Converter(log);
         var serialized = serializeToValueAsync(ImmutableMap.<String, Object>builder()
                 .put("s", "str")
                 .put("b", true)
@@ -61,7 +70,7 @@ class ComplexTypeConverterTest {
                 .put("color", "blue")
                 .build()
         ).join();
-        var data = Converter.convertValue(
+        var data = converter.convertValue(
                 "ComplexTypeConverterTest", serialized, ComplexType1.class
         );
 
@@ -89,8 +98,7 @@ class ComplexTypeConverterTest {
         public ComplexType2(
                 ComplexType1 c,
                 ImmutableList<ComplexType1> c2List,
-                ImmutableMap<String, ComplexType1> c2Map)
-        {
+                ImmutableMap<String, ComplexType1> c2Map) {
             this.c = c;
             this.c2List = c2List;
             this.c2Map = c2Map;
@@ -99,6 +107,7 @@ class ComplexTypeConverterTest {
 
     @Test
     void testTestComplexType2() {
+        var converter = new Converter(log);
         var serialized = serializeToValueAsync(ImmutableMap.<String, Object>builder()
                 .put("c", ImmutableMap.<String, Object>builder()
                         .put("s", "str1")
@@ -133,14 +142,14 @@ class ComplexTypeConverterTest {
                                 .put("d", 3.3)
                                 .put("list", List.of(true, false))
                                 .put("map", Map.of("k", 3))
-                                .put("obj", Map.of( "o", 5.5))
+                                .put("obj", Map.of("o", 5.5))
                                 .put("size", 6)
                                 .put("color", "blue")
                                 .build()
                 ))
                 .build()
         ).join();
-        var data = Converter.convertValue(
+        var data = converter.convertValue(
                 "ComplexTypeConverterTest", serialized, ComplexType2.class
         ).getValueNullable();
 
@@ -185,5 +194,40 @@ class ComplexTypeConverterTest {
         assertThat((Map<String, Object>) value.obj).containsAllEntriesOf(Map.of("o", Optional.of(5.5))); // C# didn't have Optional, it has Nullable type.
         assertThat(value.size).isEqualTo(ContainerSize.SixInch);
         assertThat(value.color).isEqualTo(ContainerColor.Blue);
+    }
+
+    @OutputCustomType
+    public static class UnexpectedNullableComplexType {
+        public final String s;
+
+        @OutputCustomType.Constructor({"s"})
+        public UnexpectedNullableComplexType(String s) {
+            this.s = s;
+        }
+    }
+
+    @Test
+    void testUnexpectedNullableComplexType() {
+        var logger = InMemoryLogger.getLogger(Level.FINEST, "ComplexTypeConverterTest#testUnexpectedNullableComplexType");
+        var inMemoryLog = DeploymentTests.mockLog(logger);
+        var converter = new Converter(inMemoryLog);
+
+        var map = new HashMap<String, Object>();
+        map.put("s", null);
+        var serialized = serializeToValueAsync(map).join();
+
+        var data = converter.convertValue(
+                "ComplexTypeConverterTest", serialized, UnexpectedNullableComplexType.class
+        ).getValueNullable();
+
+        assertThat(data).isNotNull();
+
+        var value = data.s;
+        assertThat(value).isNull();
+
+        var messages = logger.getMessages();
+        assertThat(messages).haveAtLeastOne(containsString(
+                "parameter named 's' (nr 0 starting from 0) lacks @javax.annotation.Nullable annotation, so the value is required, but there is no value to deserialize."
+        ));
     }
 }

--- a/sdk/jvm/pulumi/src/test/java/io/pulumi/serialization/internal/ConverterTests.java
+++ b/sdk/jvm/pulumi/src/test/java/io/pulumi/serialization/internal/ConverterTests.java
@@ -122,8 +122,9 @@ class ConverterTests {
         void testInvokeArgs() {
             var gson = new Gson();
             var args = new ComplexInvokeArgs1(new SimpleInvokeArgs1("value1"));
+            var converter = new Converter(log);
             serializeToValueAsync(args)
-                    .thenApply(value -> Converter.convertValue("ArgsConverterTests", value, JsonElement.class))
+                    .thenApply(value -> converter.convertValue("ArgsConverterTests", value, JsonElement.class))
                     .thenAccept(data ->
                             assertThat(data.getValueNullable())
                                     .isEqualTo(gson.fromJson("{\"v\"={\"s\"=\"value1\"}}", JsonElement.class))
@@ -134,8 +135,9 @@ class ConverterTests {
         void testResourceArgs() {
             var gson = new Gson();
             var args = new ComplexResourceArgs1(Input.of(new SimpleResourceArgs1(Input.of("value2"))));
+            var converter = new Converter(log);
             serializeToValueAsync(args)
-                    .thenApply(value -> Converter.convertValue("ArgsConverterTests", value, JsonElement.class))
+                    .thenApply(value -> converter.convertValue("ArgsConverterTests", value, JsonElement.class))
                     .thenAccept(data ->
                             assertThat(data.getValueNullable())
                                     .isEqualTo(gson.fromJson("{\"v\"={\"s\"=\"value2\"}}", JsonElement.class))
@@ -154,7 +156,8 @@ class ConverterTests {
         @Test
         void testTrue() {
             var trueValue = Value.newBuilder().setBoolValue(true).build();
-            var data = Converter.convertValue(
+            var converter = new Converter(log);
+            var data = converter.convertValue(
                     "BooleanConverterTests", trueValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isTrue();
@@ -164,7 +167,8 @@ class ConverterTests {
         @Test
         void testFalse() {
             var falseValue = Value.newBuilder().setBoolValue(false).build();
-            var data = Converter.convertValue(
+            var converter = new Converter(log);
+            var data = converter.convertValue(
                     "BooleanConverterTests", falseValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -174,7 +178,8 @@ class ConverterTests {
         @Test
         void testTrueSecret() {
             var secretValue = createSecretValue(Value.newBuilder().setBoolValue(true).build());
-            var data = Converter.convertValue(
+            var converter = new Converter(log);
+            var data = converter.convertValue(
                     "BooleanConverterTests", secretValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isTrue();
@@ -185,7 +190,8 @@ class ConverterTests {
         @Test
         void testFalseSecret() {
             var secretValue = createSecretValue(Value.newBuilder().setBoolValue(false).build());
-            var data = Converter.convertValue(
+            var converter = new Converter(log);
+            var data = converter.convertValue(
                     "BooleanConverterTests", secretValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -196,8 +202,9 @@ class ConverterTests {
         @Test
         void testNonBooleanThrows() {
             var wrongValue = Value.newBuilder().setStringValue("").build();
+            var converter = new Converter(log);
             assertThatThrownBy(
-                    () -> Converter.convertValue("BooleanConverterTests", wrongValue, Boolean.class)
+                    () -> converter.convertValue("BooleanConverterTests", wrongValue, Boolean.class)
             ).isInstanceOf(UnsupportedOperationException.class)
                     .hasMessageContaining("Expected 'java.lang.Boolean' but got 'java.lang.String' while deserializing");
         }
@@ -209,8 +216,9 @@ class ConverterTests {
                 .setOptions(new TestOptions(true))
                 .setMockGlobalInstance();
 
+            var converter = new Converter(log);
             var nullValue = Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", nullValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -224,8 +232,9 @@ class ConverterTests {
                 .setOptions(new TestOptions(false))
                 .setMockGlobalInstance();
 
+            var converter = new Converter(log);
             var nullValue = Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", nullValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -234,8 +243,9 @@ class ConverterTests {
 
         @Test
         void testUnknownProducesFalseUnknown() {
+            var converter = new Converter(log);
             var unknownValue = Value.newBuilder().setStringValue(Constants.UnknownValue).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", unknownValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -244,9 +254,10 @@ class ConverterTests {
 
         @Test
         void testEmptyStringThrows() {
+            var converter = new Converter(log);
             var wrongValue = Value.newBuilder().setStringValue("").build();
             assertThatThrownBy(
-                    () -> Converter.convertValue("BooleanConverterTests", wrongValue, Boolean.class)
+                    () -> converter.convertValue("BooleanConverterTests", wrongValue, Boolean.class)
             ).isInstanceOf(UnsupportedOperationException.class)
                     .hasMessageContaining("Expected 'java.lang.Boolean' but got 'java.lang.String' while deserializing");
 
@@ -254,8 +265,9 @@ class ConverterTests {
 
         @Test
         void testOptionalTrueWrapped() {
+            var converter = new Converter(log);
             var trueValue = Value.newBuilder().setBoolValue(true).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", trueValue, TypeShape.optional(Boolean.class)
             );
             assertThat(data.getValueNullable()).isNotNull().isPresent().contains(true);
@@ -264,8 +276,9 @@ class ConverterTests {
 
         @Test
         void testOptionalTrueUnwrapped() {
+            var converter = new Converter(log);
             var trueValue = Value.newBuilder().setBoolValue(true).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", trueValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isTrue();
@@ -274,8 +287,9 @@ class ConverterTests {
 
         @Test
         void testOptionalFalseWrapped() {
+            var converter = new Converter(log);
             var trueValue = Value.newBuilder().setBoolValue(false).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", trueValue, TypeShape.optional(Boolean.class)
             );
             assertThat(data.getValueNullable()).isNotNull().isPresent().contains(false);
@@ -284,8 +298,9 @@ class ConverterTests {
 
         @Test
         void testOptionalFalseUnwrapped() {
+            var converter = new Converter(log);
             var trueValue = Value.newBuilder().setBoolValue(false).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", trueValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isNotNull().isFalse();
@@ -294,8 +309,9 @@ class ConverterTests {
 
         @Test
         void testNullToOptionalEmpty() {
+            var converter = new Converter(log);
             var nullValue = Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", nullValue, TypeShape.optional(Boolean.class)
             );
             assertThat(data.getValueNullable()).isNotNull().isEmpty();
@@ -304,8 +320,9 @@ class ConverterTests {
 
         @Test
         void testNullToNullUnwrapped() {
+            var converter = new Converter(log);
             var nullValue = Value.newBuilder().setNullValue(NullValue.NULL_VALUE).build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "BooleanConverterTests", nullValue, Boolean.class
             );
             assertThat(data.getValueNullable()).isFalse();
@@ -319,8 +336,9 @@ class ConverterTests {
         @ParameterizedTest
         @EnumSource(ContainerSize.class)
         void testCustomIntEnum(ContainerSize input) {
+            var converter = new Converter(log);
             serializeToValueAsync(input)
-                    .thenApply(value -> Converter.convertValue("EnumConverterTests", value, ContainerSize.class))
+                    .thenApply(value -> converter.convertValue("EnumConverterTests", value, ContainerSize.class))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable()).isNotNull();
                         assertThat(input.asDouble()).isEqualTo(data.getValueNullable().asDouble());
@@ -331,8 +349,9 @@ class ConverterTests {
         @ParameterizedTest
         @EnumSource(JarSize.class)
         void testStandardIntEnum(JarSize input) {
+            var converter = new Converter(log);
             serializeToValueAsync(input)
-                    .thenApply(value -> Converter.convertValue("EnumConverterTests", value, JarSize.class))
+                    .thenApply(value -> converter.convertValue("EnumConverterTests", value, JarSize.class))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable()).isNotNull();
                         assertThat(input.asDouble()).isEqualTo(data.getValueNullable().asDouble());
@@ -343,8 +362,9 @@ class ConverterTests {
         @ParameterizedTest
         @EnumSource(ContainerColor.class)
         void testCustomStringEnum(ContainerColor input) {
+            var converter = new Converter(log);
             serializeToValueAsync(input)
-                    .thenApply(value -> Converter.convertValue("EnumConverterTests", value, ContainerColor.class))
+                    .thenApply(value -> converter.convertValue("EnumConverterTests", value, ContainerColor.class))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable()).isNotNull();
                         assertThat(input.getValue()).isEqualTo(data.getValueNullable().getValue());
@@ -355,8 +375,9 @@ class ConverterTests {
         @ParameterizedTest
         @EnumSource(ContainerBrightness.class)
         void testCustomDoubleEnum(ContainerBrightness input) {
+            var converter = new Converter(log);
             serializeToValueAsync(input)
-                    .thenApply(value -> Converter.convertValue("EnumConverterTests", value, ContainerBrightness.class))
+                    .thenApply(value -> converter.convertValue("EnumConverterTests", value, ContainerBrightness.class))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable()).isNotNull();
                         assertThat(input.getValue()).isEqualTo(data.getValueNullable().getValue());
@@ -377,7 +398,8 @@ class ConverterTests {
         @ParameterizedTest
         @MethodSource("io.pulumi.serialization.internal.ConverterTests#testConvertingNonconvertibleValuesThrows")
         void testConvertingNonconvertibleValuesThrows(Class<?> targetType, Value value) {
-            assertThatThrownBy(() -> Converter.convertValue("EnumConverterTests", value, targetType))
+            var converter = new Converter(log);
+            assertThatThrownBy(() -> converter.convertValue("EnumConverterTests", value, targetType))
                     .isInstanceOf(UnsupportedOperationException.class)
                     .hasMessageContaining("Expected value that match any of enum");
         }
@@ -388,7 +410,8 @@ class ConverterTests {
 
         @Test
         void testLeft() {
-            var data = Converter.convertValue(
+            var converter = new Converter(log);
+            var data = converter.convertValue(
                     "EitherConverterTests", Value.newBuilder().setNumberValue(1).build(), TypeShape.either(Integer.class, String.class));
             assertThat(data.isKnown()).isTrue();
             assertThat(data.getValueNullable()).isNotNull();
@@ -398,8 +421,9 @@ class ConverterTests {
 
         @Test
         void testRight() {
+            var converter = new Converter(log);
             var value = Value.newBuilder().setStringValue("foo").build();
-            var data = Converter.convertValue(
+            var data = converter.convertValue(
                     "EitherConverterTests", value, TypeShape.either(Integer.class, String.class));
             assertThat(data.isKnown()).isTrue();
             assertThat(data.getValueNullable()).isNotNull();
@@ -413,6 +437,7 @@ class ConverterTests {
 
         @Test
         public void ignoreInternalProperty() {
+            var converter = new Converter(log);
             var value = Value.newBuilder().setStructValue(
                     Struct.newBuilder()
                             .putFields("a", Value.newBuilder().setStringValue("b").build())
@@ -423,7 +448,7 @@ class ConverterTests {
                     .addParameter(String.class)
                     .addParameter(String.class)
                     .build();
-            var data = Converter.convertValue("InternalPropertyTests", value, shape);
+            var data = converter.convertValue("InternalPropertyTests", value, shape);
             assertThat(data.isKnown()).isTrue();
             //noinspection unchecked
             assertThat(data.getValueNullable()).isNotNull();
@@ -443,10 +468,11 @@ class ConverterTests {
         @ParameterizedTest
         @MethodSource("io.pulumi.serialization.internal.ConverterTests#testJsons")
         void testJsons(String json, String expected) {
+            var converter = new Converter(log);
             var element = gson.fromJson(json, JsonElement.class);
             serializeToValueAsync(element)
                     .thenAccept(serialized -> {
-                        var converted = Converter.convertValue("JsonConverterTests", serialized, JsonElement.class);
+                        var converted = converter.convertValue("JsonConverterTests", serialized, JsonElement.class);
                         assertThat(converted.getValueNullable()).isNotNull();
                         assertThat(converted.getValueNullable().toString()).isEqualTo(expected);
                     })
@@ -459,9 +485,10 @@ class ConverterTests {
 
         @Test
         void testEmptyList() {
+            var converter = new Converter(log);
             List<Boolean> emptyList = List.of();
             serializeToValueAsync(emptyList)
-                    .thenApply(value -> Converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
+                    .thenApply(value -> converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable()).isNotNull().isEmpty();
                         assertThat(data.isKnown()).isTrue();
@@ -470,9 +497,10 @@ class ConverterTests {
 
         @Test
         void testListWithElement() {
+            var converter = new Converter(log);
             var listWithElement = List.of(true);
             serializeToValueAsync(listWithElement)
-                    .thenApply(value -> Converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
+                    .thenApply(value -> converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable())
                                 .isNotNull()
@@ -484,9 +512,10 @@ class ConverterTests {
 
         @Test
         void testSecretListWithElement() {
+            var converter = new Converter(log);
             var secretListWithElement = Output.ofSecret(List.of(true));
             serializeToValueAsync(secretListWithElement)
-                    .thenApply(value -> Converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
+                    .thenApply(value -> converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable())
                                 .isNotNull()
@@ -500,9 +529,10 @@ class ConverterTests {
 
         @Test
         void testListWithSecretElement() {
+            var converter = new Converter(log);
             var listWithSecretElement = List.of(Output.ofSecret(true));
             serializeToValueAsync(listWithSecretElement)
-                    .thenApply(value -> Converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
+                    .thenApply(value -> converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable())
                                 .isNotNull()
@@ -515,9 +545,10 @@ class ConverterTests {
 
         @Test
         void testListWithUnknownElement() {
+            var converter = new Converter(log);
             var listWithUnknownElement = List.of(InputOutputTests.unknown(true));
             serializeToValueAsync(listWithUnknownElement)
-                    .thenApply(value -> Converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
+                    .thenApply(value -> converter.convertValue("ListConverterTests", value, TypeShape.list(Boolean.class)))
                     .thenAccept(data -> {
                         assertThat(data.getValueNullable())
                                 .isNotNull()
@@ -534,6 +565,7 @@ class ConverterTests {
 
         @Test
         void testSimpleCase() {
+            var converter = new Converter(log);
             var value = Value.newBuilder()
                     .setStructValue(
                             Struct.newBuilder()
@@ -553,7 +585,7 @@ class ConverterTests {
                                             ).build()
                                     ).build()
                     ).build();
-            var data = Converter.convertValue("", value, RecursiveType.class);
+            var data = converter.convertValue("", value, RecursiveType.class);
 
             assertThat(data.isKnown()).isTrue();
             assertThat(data.getValueNullable()).isNotNull();

